scrapy/core/spidermw.py

```diff
@@ -82,11 +110,22 @@ def _process_spider_exception(self, response: Response, spider: Spider, _failure
         for method_index, method in enumerate(method_list, start=start_index):
             if method is None:
                 continue
+            method = cast(Callable, method)
             result = method(response=response, exception=exception, spider=spider)
             if _isiterable(result):
                 # stop exception handling by handing control over to the
                 # process_spider_output chain if an iterable has been returned
-                return self._process_spider_output(response, spider, result, method_index + 1)
+                dfd: Deferred = self._process_spider_output(response, spider, result, method_index + 1)
+                # _process_spider_output() returns a Deferred only because of downgrading so this can be
+                # simplified when downgrading is removed.
+                if dfd.called:
+                    # the result is available immediately if _process_spider_output didn't do downgrading
+                    return dfd.result
+                else:
+                    # we forbid waiting here because otherwise we would need to return a deferred from
+                    # _process_spider_exception too, which complicates the architecture
+                    msg = f"Async iterable returned from {method.__qualname__} cannot be downgraded"
+                    raise _InvalidOutput(msg)
```

Comment:
As I was working on a [proposal to refactor the new documentation](https://github.com/wRAR/scrapy/pull/1), I got stuck at trying to cover `process_spider_exception` properly.

I was trying to understand in which scenarios trigger this exception, and I came to realize the following about `process_spider_exception`:

- If defined as an asynchronous generator, it cannot return `None`. If it does not yield anything, that will count as an empty iterable. So with `async def` + `yield` we lose the ability to get Scrapy to call the next `process_spider_exception` method.
- If defined as a coroutine, an exception is raised because it is neither an iterable nor `None` (it is a coroutine), so `async def` _without_ `yield` is not an option either.

In other words, `async def process_spider_exception` can never return `None` to get Scrapy to let the next `process_spider_exception` work.

I wonder if it would not be better to limit `process_spider_exception` to being synchronous for now, until we can figure out an API for it that allows the `None` behavior as well. This feels similar to start_requests, i.e. we may need to define some special object that may be yielded from the asynchronous generator to indicate the same as `None` does in the synchronous version.